runtime/domain.c

```diff
@@ -389,6 +389,24 @@ int caml_reallocate_minor_heap(asize_t wsize)
   return 0;
 }
 
+/* This variable is owned by [all_domains_lock]. */
+static uintnat next_domain_unique_id = 0;
+
+/* Precondition: you must own [all_domains_lock]. */
+static uintnat fresh_domain_unique_id(void) {
+    uintnat next = next_domain_unique_id++;
+
+    /* On 32-bit systems, there is a risk of wraparound of the unique
+       id counter. We have decided to let that happen and live with
+       it, but we still ensure that id 0 is not reused, to avoid
+       having new domains believe that they are the main domain. */
+    if (next_domain_unique_id == 0)
+      next_domain_unique_id++;
+
+    return next;
+}
```

Comment:
You're doing a post-increment, so the first time `fresh_domain_unique_id` is called  it will return... 0 !

Either you initialize `next_domain_unique_id` with 1 or you write this:
```suggestion
static uintnat fresh_domain_unique_id(void) {
    next_domain_unique_id++;

    /* On 32-bit systems, there is a risk of wraparound of the unique
       id counter. We have decided to let that happen and live with
       it, but we still ensure that id 0 is not reused, to avoid
       having new domains believe that they are the main domain. */
    if (next_domain_unique_id == 0) next_domain_unique_id = 1;

    return next_domain_unique_id;
}
```

Reply:
This is by design: the first domain is initialize by a call to `create_domain` in `caml_init_domains`, and I want this very first call to return 0.

https://github.com/ocaml/ocaml/blob/6a7a7b36fa2b6f4c4f3d210c8ecaff6faa094adf/runtime/domain.c#L618-L619

I guess that this should be made more explicit, possibly by doing an assert on `domain_self`'s unique id right after initialization.

Reply:
Ouch... that's a fragile design!  You'd better document it, then.

Reply:
This is documented now and tested by a `CAMLassert` and in the testsuite.

Reply:
(I'm secretely hoping that the "Changes requested" status will be changed back to "merge merge merge!".)

Reply:
Thanks for the added comments.  Will remove the mark of infamy ASAP.